### ISSUE TITLE
logstash benchmark?

### ISSUE THREAD

**Issue by fandrei** [elastic/logstash#2793]
The latest changelog says:

```
Performance improvements: Logstash 1.5.0 is much faster -- we have improved the throughput 
    of grok filter in some cases by 100%. In our benchmark testing, using only grok filter and
    ingesting apache logs, throughput increased from 34K eps to 50K eps. 
```

Where can I find this benchmark?

**Comment by purbon** [comment]
For now this is not open, but it should be. This is a short term objective for us.

For if you want to reproduce this you can basically use a simple config like:

```
input {
	stdin { type => "apache" }
}

filter {
	grok {
		match => {"message" => "%{COMBINEDAPACHELOG}"}
	}
}

output {
	stdout { codec => line }
}
```

with a bunch of apache logs and compute the eps. 

Another way of getting this date too is to use the experimental gem you can find at https://github.com/elastic/logstash-integration-testing/, this was has the same code base as the one we used to run the benchmarks.

This last gem is in an experimental code base, but feel free to use it and ask me questions if you've troubles.

Do you think this provide you all the necessary information for you, can we close this issue ?

**Comment by fandrei** [comment]
ok

**Comment by purbon** [comment]
Thanks @fandrei, let me know if you find interesting benchmarking questions.

**Comment by fandrei** [comment]
@purbon, I wonder what hardware are you using for your benchmarks?

**Comment by purbon** [comment]
Where where using a big machine.

31 Gb RAM
8 core Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
...

but the way logstash was run was really using the defaults in terms of workers, java_opts, etc.

the java vm was

```
java version "1.7.0_65"
OpenJDK Runtime Environment (IcedTea 2.5.3) (7u71-2.5.3-0ubuntu0.14.04.1)
OpenJDK 64-Bit Server VM (build 24.65-b04, mixed mode)
```

**Comment by fandrei** [comment]
@purbon, thanks

**Comment by purbon** [comment]
however you should also be gettings the similar numbers with smaller configurations.

**Comment by fandrei** [comment]
@purbon, I'm getting ~12k log events with empty filter (logstash v1.5), what is quite different from your result.
I'm running it on Windows, though. i5-3550, 16Gb RAM, Java 8 update 40.
USE_JRUBY=1

**Comment by purbon** [comment]
This test where done in a unix machine, I should have said that ... but your point about windows is nice. BTW in 1.5 you don't really need to use the USE_JRUBY variable.

can you run in windows the same with 1.4 ? would be very interesting to see the comparison.

can you also share here your configuration file?

**Comment by fandrei** [comment]
@purbon, the same test gives ~10k when using logstash 1.4.2
My test config is quite simple: 

```
input {
  stdin {
  }
}

filter {
} 

output {
  stdout {
  }
}
```

PS Java VM installed is 32 bit

**Comment by purbon** [comment]
@fandrei Interesting and something to be analysed, have you tried what happened with a sample grok filter in windows? 

config like

```
input {
    stdin { type => "apache" }
}

filter {
    grok {
        match => {"message" => "%{COMBINEDAPACHELOG}"}
    }
}

output {
    stdout { codec => line }
}
```

**Comment by fandrei** [comment]
@purbon, not yet. Do you have any specific apache input example you are using?

**Comment by purbon** [comment]
you can use https://github.com/elastic/logstash-integration-testing/blob/master/examples/input/apache_log.txt and generate as many log lines as you want.

**Comment by fandrei** [comment]
@purbon, ~5700 events per second using logstash 1.5

**Comment by fajabird** [comment]
I would like to share my benchmark results from our production system. Setup is like this:

node1: ES 1.7 + LS (logs received via tcp)
node2: ES 1.7 + LS (logs reveived via syslog)
node3: ES 1.7 (query) + Kibana 3.1 and 4.0

# Shards: 2
# Replicas: 0

Benchmark via metric-filter (for each node seperately - here only node1 matters as this has much higher load). My pileline for node1 looks like this:

input { tcp { port => 8080 tags => "tcp"  } }

output {  elasticsearch {  host => "ip-of-node1"  protocol => "transport"   }  }

filter {  metrics {  meter => "events"   add_tag => "metric"  } }

# the work horse is grok, triggered by a condition (host-ip is matching the regex):

filter { if  [host] =~ /^aaa.bbb.ccc.(40|50|52|54|56)(:\d+)?/ {
    grok {
      add_tag => "proxysg"
      add_field => { "logsource" => "PROXYSG" }
      match => [ "message", "%{PROXYSG}" ]
      patterns_dir => "/etc/logstash/conf.d/patterns"
    }
    date {
      match => [ "date", "YYYY-MM-dd HH:mm:ss" ]
      timezone => "UTC"
    }
} }

1. Logstash 1.5.6 + ES 1.7.4
1 filter worker (-w 1): 800 EPS
4 filter workers (-w 4): 2000 EPS

2. Logstash 1.4.5-1-a2bacae + ES 1.7.4
1 filter worker (-w 1): 2200 EPS
4 filter workers (-w 4): 3400 EPS

BTW: I also tried ES 2.1 + LS 2.1 which have me even lower EPS rates like 1200 for 4 filter workers. So I downgraded to 1.5 and noch I'm back at 1.4 because of the performance boost. I would be glad to support you with more information if needed. BTW. Logstash config is always the exact same. Tried different LS_HEAP_SIZE with nearly no effect as well as LS_NICE levels.

**Comment by suyograo** [comment]
@fandrei thanks, those numbers are concerning. could you also provide a bunch of sample logs for this configuration? You can replace any sensitive information..

**Comment by fajabird** [comment]
ProxySG logs look like this:

#Fields: date time c-ip cs-username cs-auth-group sc-status s-action sc-bytes cs-bytes cs-method cs-uri-scheme cs-host cs-uri-port cs-uri-path cs(Referer) cs-categories

2016-01-19 01:01:30 aaa.bbb.ccc.56 - - 200 TCP_TUNNELED 1148 1614 CONNECT tcp apis.google.com 443 / - "Allowed Sites;Non-Viewable/Infrastructure"
2016-01-19 01:01:30 aaa.bbb.ccc.29 - - 200 TCP_TUNNELED 9232 2131 CONNECT tcp safebrowsing-cache.google.com 443 / - "Allowed Sites;Non-Viewable/Infrastructure"
2016-01-19 01:01:30 aaa.bbb.ccc.38 - - 200 TCP_TUNNELED 5579 4092 CONNECT tcp mail.google.com 443 / - "Allowed Sites;Email"
2016-01-19 01:01:31 aaa.bbb.ccc.14 - - 200 TCP_TUNNELED 870 2059 CONNECT tcp play.google.com 443 / - "Allowed Sites;Shopping;Software Downloads"
2016-01-19 01:01:31 aaa.bbb.ccc.56 username1 GROUP\Users 302 TCP_POLICY_REDIRECT 898 555 GET http www.googletagservices.com 80 /tag/js/gpt.js http://www.hurriyet.com.tr/ "Web Ads/Analytics"
2016-01-19 01:01:31 aaa.bbb.ccc.56 username1 GROUP\Users 304 TCP_HIT 548 1724 GET http s.hurriyet.com.tr 80 /static/images/hurriyet/follow_googleplus.png http://www.hurriyet.com.tr/ "News/Media"
2016-01-19 01:01:31 aaa.bbb.ccc.56 username1 GROUP\Users 302 TCP_POLICY_REDIRECT 898 432 GET http www.google-analytics.com 80 /analytics.js http://www.hurriyet.com.tr/ "Web Ads/Analytics"
2016-01-19 01:01:32 aaa.bbb.ccc.56 username1 GROUP\Users 304 TCP_HIT 553 1720 GET http s.hurriyet.com.tr 80 /static/images/hurriyet/footer_google.png http://www.hurriyet.com.tr/ "News/Media"
2016-01-19 01:01:32 aaa.bbb.ccc.56 username1 GROUP\Users 304 TCP_HIT 590 533 GET http ajax.googleapis.com 80 /ajax/libs/jquery/1.11.1/jquery.min.js http://www.hurriyet.com.tr/ "Technology/Internet"
2016-01-19 01:01:32 aaa.bbb.ccc.146 username2 GROUP\Users 200 TCP_NC_MISS 37862 483 GET http www.googletagmanager.com 80 /gtm.js http://www.hurriyet.com.tr/ "Technology/Internet"

**Comment by fajabird** [comment]
my grok parser regex:

PROXYSG %{TIMESTAMP_ISO8601:date} %{DATA:c_ip} %{DATA:cs_username} %{DATA:cs_auth_group} %{INT:sc_status} %{DATA:s_action} %{INT:sc_bytes:int} %{INT:cs_bytes:int} %{DATA:cs_method} %{DATA:cs_uri_scheme} %{DATA:cs_host} %{INT:cs_uri_port} %{DATA:cs_uri_path} %{DATA:cs_referer} "%{DATA:cs_categories}"

**Comment by fajabird** [comment]
I changed my pipeline on node1 a bit. I deleted all other conditions which trigger different grok parsers for different sources (this way I can use the same filter configuration across all nodes). Now node1 only has TCP-input, ES-Output and a single grok filter only for ProxySG-logs. I was glad to see that this optimization had only minimal impact on performance (3600EPS instead 3400EPS) - so the conditions inside my filter configuration seem ok. Now I wanted to see how the two main ES protocols do:

Again: Logstash 1.4.5-1-a2bacae + ES 1.7.4 with 4 filter workers (-w 4):

protocol=transport: 3600 EPS
protocol=http: 3500 EPS

**Comment by fajabird** [comment]
Based on this optimized pipeline I updated LS to 1.5.6 without changing any configuration files (same startup parameters, same LS pipeline) - and I was supprised:

Logstash 1.5.6-1 + ES 1.7.4 with 4 filter workers (-w 4):

protocol=transport: 4000 EPS
protocol=http: 3500 EPS

I must admit that results vary a lot. And while I was happy to see 4000 EPS I was only able to get this result once - mostly it settled at around 3600 EPS. So maybe LS 1.4.5 would be able to reach 4000 EPS as well.

After this result I updated LS to 2.1 again without changing anything expect now only http-protocol is supported:

protocol=http: 3200 EPS. Still much higher than what I had in my first test.

I conclude that with the optimized pipeline the EPS figures are very close - still I see a small performance drop with LS 2.1

**Comment by fajabird** [comment]
Next I wanted to find out why my first results were much lower and I did. 

LS 1.5.6: 

old pipeline with dnsmasq-lookups: 1800 EPS
old pipeline without DNS-filter: 3400 EPS

I conclude that LS 1.5.6 suffers a lot if dns filter is beeing used although top reports only 5% cpuload on a single core by dnsmasq.

With LS 1.4 performance drop by DNS is much lower or not existing at all. 

old pipeline with dnsmasq-lookups: 3200 EPS
old pipeline without DNS-filter: 3300 EPS

**Comment by untergeek** [comment]
I wonder if using an LRU cache in the DNS filter (as with the geoip and useragent filters) would improve performance here?
